File: code/src/graph.cpp
#include "graph.hpp"
#include <iostream>

namespace graph
{
    BundledEdge::BundledEdge() : color(std::numeric_limits<size_t>::max()) {}
    BundledEdge::BundledEdge(size_t _color) : color(_color) {}

    BundledEdge2::BundledEdge2(Edge *_id) : id(_id) {}

    size_t color(const Graph &GG, Edge &edge)
    {
        return GG[edge].color;
    }

    // TODO: otimizar usando referencia de edges de G[i] para GG
    std::pair<bool, graph::Edge> checkEdge(int u, int v, size_t color, const Instance &instance)
    {
        auto &[GG, G] = instance;
        auto ans = edge(u, v, G[color]);
        if (!ans.second)
            return {false, Edge()};
        auto EDGE = G[color][ans.first].id;
        std::cout << "ANS: " << ans.first << std::endl;
        std::cout << "EDGE: " << *EDGE << std::endl;
        std::cout << "COLOR: " << (*GG)[*EDGE].color << std::endl;
        return {true, *EDGE};
        for (auto edge : boost::make_iterator_range(boost::out_edges(u, *GG)))
        {
            if (target(edge, *GG) == v && (*GG)[edge].color == color)
            {
                return {true, edge};
            }
        }
        throw std::runtime_error("Should not reach here. Edge not found on big graph.");
    };

    Path::Path(graphPointer _G) : G(_G) {}
    Path::Path(graphPointer _G, const std::vector<Vertex> &_vertices, const std::vector<Edge> &_edges) : G(_G), vertices(_vertices), edges(_edges)
    {
        assert(vertices.size() == edges.size() + 1);
        std::vector<int> usedVertex(boost::num_vertices(*G), 0), usedColors(boost::num_vertices(*G), 0);
        for (const auto &vertex : vertices)
        {
            if (usedVertex[vertex] > 0)
            {
                throw std::runtime_error("Path has repeated vertices.");
            }
            usedVertex[vertex]++;
        }
        for (const auto &edge : edges)
        {
            if (usedColors[(*G)[edge].color] > 0)
            {
                throw std::runtime_error("Path has repeated colors.");
            }
            usedColors[(*G)[edge].color]++;
        }
        for (int i = 0; i + 1 < (int)vertices.size(); i++)
        {
            bool valid = false;
            if (source(edges[i], *G) != (size_t)vertices[i] && target(edges[i], *G) != (size_t)vertices[(i + 1) % vertices.size()])
                valid = true;
            if (source(edges[i], *G) != (size_t)vertices[(i + 1) % vertices.size()] && target(edges[i], *G) != (size_t)vertices[i])
                valid = true;
            if (!valid)
            {
                throw std::runtime_error("Cycle has invalid edges.");
            }
        }
    }

    size_t Path::size() const
    {
        return edges.size();
    }

    Vertex Path::back() const
    {
        return vertices.back();
    }

    // returns false if the edge does not exist
    // TODO: FIX ADD CORRECT DESCRIPTOR
    bool Path::push_back(Vertex v, Edge e = Edge())
    {
        if (vertices.size() > 0)
        {
            assert(e != Edge() && "Edge must be provided.");
            edges.push_back(e);
        }
        vertices.push_back(v);
        return true;
    }

    void Path::pop_back()
    {
        assert(vertices.size() > 0 && "Cannot pop from empty path.");
        vertices.pop_back();
        edges.pop_back();
    }

    Cycle::Cycle(graphPointer _G) : G(_G) {}

    Cycle::Cycle(graphPointer _G, const std::vector<Vertex> &_vertices, const std::vector<Edge> &_edges) : G(_G), vertices(_vertices), edges(_edges)
    {
        assert(vertices.size() == edges.size());
        std::vector<int> usedVertex(boost::num_vertices(*G), 0), usedColors(boost::num_vertices(*G), 0);
        for (const auto &vertex : vertices)
        {
            if (usedVertex[vertex] > 0)
            {
                throw std::runtime_error("Cycle has repeated vertices.");
            }
            usedVertex[vertex]++;
        }
        for (const auto &edge : edges)
        {
            if (usedColors[(*G)[edge].color] > 0)
            {
                throw std::runtime_error("Cycle has repeated colors.");
            }
            usedColors[(*G)[edge].color]++;
        }
        for (int i = 0; i < (int)vertices.size(); i++)
        {
            bool valid = false;
            if (source(edges[i], *G) != (size_t)vertices[i] && target(edges[i], *G) != (size_t)vertices[(i + 1) % vertices.size()])
                valid = true;
            if (source(edges[i], *G) != (size_t)vertices[(i + 1) % vertices.size()] && target(edges[i], *G) != (size_t)vertices[i])
                valid = true;
            if (!valid)
            {
                throw std::runtime_error("Cycle has invalid edges.");
            }
        }
    }

    size_t Cycle::size() const
    {
        return edges.size();
    }
}

File: python/src/graph.py
import sys
from typing import Dict, List, Optional

type Color = int
# Definindo Edge e Vertex como alias de int
type Vertex = int

class Edge:
    source: Vertex
    target: Vertex
    color: Color
    def __init__(self, source: Vertex, target: Vertex, color: Color):
        self.source = source
        self.target = target
        self.color = color

# Estrutura auxiliar para Graph
class BundledEdge:
    def __init__(self, color: int = sys.maxsize):
        self.color = color

class Graph:
    n: int
    edges: List[Edge]
    G: List[Dict]
    adjacency: List[List[List[Vertex]]]
    def __init__(self, n: int):
        self.n = n
        self.edges = []
        self.G = [{} for _ in range(n)]
        self.adjacency = [[[] for _ in range(n)] for _ in range(n)]  

    def add_edge(self, u: int, v: int, color: int):
        self.edges.append(Edge(u, v, color))
        self.G[color][(u, v)] = True
        self.G[color][(v, u)] = True
        self.adjacency[color][u].append(v)
        self.adjacency[color][v].append(u)

    def get_edges(self, color: Color) -> List[Edge]:
        return [edge for edge in self.edges if edge.color == color]

    def get_incident_edges(self, color: Color, u: Vertex) -> List[Edge]:
        return [edge for edge in self.edges if edge.color == color and (edge.source == u or edge.target == u)]
    
    def check_edge(self, u: Vertex, v: Vertex, color: Color) -> Optional[Edge]:
        return Edge(u, v, color) if (u, v) in self.G[color] else None

    def get_edge(self, u: int, v: int, color: int) -> Edge:
        edge = self.check_edge(u, v, color)
        assert edge is not None
        return edge

class Path:
    vertices: List[Vertex]
    edges: List[Edge]
    def __init__(self, G: Graph, vertices: Optional[List[Vertex]] = None, edges: Optional[List[Edge]] = None):
        self.G = G
        n = G.n
        self.vertices: List[Vertex] = vertices if vertices else []
        self.edges: List[Edge] = edges if edges else []
        if vertices and edges:
            assert len(vertices) == len(edges) + 1
            used_vertex = [0] * n
            used_colors = [0] * n
            for vertex in vertices:
                if used_vertex[vertex] > 0:
                    raise RuntimeError("Path has repeated vertices.")
                used_vertex[vertex] += 1
            for edge in edges:
                if used_colors[edge.color] > 0:
                    raise RuntimeError("Path has repeated colors.")
                used_colors[edge.color] += 1
            for i in range(len(vertices) - 1):
                valid = False
                if edges[i].source != vertices[i] and edges[i].target != vertices[(i + 1) % len(vertices)]:
                    valid = True
                if edges[i].source != vertices[(i + 1) % len(vertices)] and edges[i].target != vertices[i]:
                    valid = True
                if not valid:
                    raise RuntimeError("Cycle has invalid edges.")

    def size(self) -> int:
        return len(self.edges)

    def back(self) -> Vertex:
        return self.vertices[-1]

    def push_back(self, v: Vertex, e: Optional[Edge] = None) -> bool:
        if len(self.vertices) > 0:
            assert e is not None, "Edge must be provided."
            self.edges.append(e)
        self.vertices.append(v)
        return True

    def pop_back(self):
        assert len(self.vertices) > 0, "Cannot pop from empty path."
        self.vertices.pop()
        self.edges.pop()

class Cycle:
    def __init__(self, G: Graph, vertices: Optional[List[Vertex]] = None, edges: Optional[List[Edge]] = None):
        self.G = G
        n = G.n
        self.vertices = vertices if vertices else []
        self.edges = edges if edges else []
        if vertices and edges:
            assert len(vertices) == len(edges)
            used_vertex = [0] * n
            used_colors = [0] * n
            for vertex in vertices:
                if used_vertex[vertex] > 0:
                    raise RuntimeError("Cycle has repeated vertices.")
                used_vertex[vertex] += 1
            for edge in edges:
                if used_colors[edge.color] > 0:
                    raise RuntimeError("Cycle has repeated colors.")
                used_colors[edge.color] += 1
            for i in range(len(vertices)):
                if edges[i].source != vertices[i]:
                    edges[i].source, edges[i].target = edges[i].target, edges[i].source
                if edges[i].source != vertices[i] or edges[i].target != vertices[(i + 1) % len(vertices)]:
                    raise RuntimeError("Cycle has invalid edges.")

    def size(self) -> int:
        return len(self.edges)
